src/man/tta.7
.TH tta 7 2024-04-28 "version v1.2.2" AALTITOAD

.SH NAME
tta \- tick tock automata introduction

.SH THE BASICS
TTA's (Tick Tock Automata) are an Automata based theory that can model business logic for all kinds of automation systems. A TTA consists of a set of locations and transitions between them. Each transition can be guarded by a boolean expression and have a set of variable assignments as a "consequence" of taking the transition.

A good analogy to TTA is a statemachine, where a system can transition from one state to another. Below is an example of a simple TTA that controls a light based on a button input.

        \fBbtn_is_pressed  light\fR := \fItrue\fR
      ___  ----------------------->  ___
 --> |OFF|                          |ON |
     `---´ <----------------------- `---´
        \fBlight\fR := \fIfalse\fR  \fB!btn_is_pressed\fR

Starting in the OFF location, there's only one outgoing edge with the guard that checks if \fBbtn_is_pressed\fR is \fItrue\fR. If it is, we move to the ON location and set the variable \fBlight\fR to \fItrue\fR as a consequence of taking the edge/transition. Then, when the button is released we return to the OFF locaiton and \fBlight\fR is set to \fIfalse\fR again.

What makes TTAs different from any other transition system is that the \fBbtn_is_pressed\fR variable is an external variable, meaning that it cannot be assigned to any value by the transition system, it can only change based on an external input. External inputs are read when we are not taking transitions. This split of the semantics is where the name Tick Tock Automata comes from:

.RS
.IP \fITick-step\fP
evaluate guards, apply updates, change locations
.IP \fITock-step\fP
read external inputs, write external outputs
.RE

Taking the syntax one step further, you can have many of these TTAs running in parallel, sharing the same pool of internal and external variables. Such a network is called simply a network of tick tock automata (NTTA).

.SH FORMALISATION
For a more detailed formal mathematical description of TTAs and networks
of TTAs, please refer to the aaltitoad related papers.

.SH AUTHOR
Asger Gitz\-Johansen <asger.gitz@hotmail.com>.

.SH COPYRIGHT
            Copyright (C) 2024  Asger Gitz-Johansen

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>.

.SH SEE ALSO
simulator(1), verifier(1)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v0.11.0](https://github.com/sillydan1/AALTITOAD/releases/tag/v0.11.0) - 2022-09-19

<small>[Compare with v0.10.2](https://github.com/sillydan1/AALTITOAD/compare/v0.10.2...v0.11.0)</small>

### Bug Fixes
- Forgot to update fischer-10 ([3509839](https://github.com/sillydan1/AALTITOAD/commit/3509839dd5de270fa9196ac17382c5b21ccece21) by Asger Gitz-Johansen).
- Interesting tocker will now actually handle clock delay stuff ([dde1f65](https://github.com/sillydan1/AALTITOAD/commit/dde1f65200fd37f04c899aba1e55604bf7b69a5d) by Asger Gitz-Johansen).
- Fischer-2/5 test sets should reset their timer to 0_ms instead of 0 ([23ce8c6](https://github.com/sillydan1/AALTITOAD/commit/23ce8c61d357d27497079e36d11e01f060ef6c60) by Asger Gitz-Johansen).
- Fix compilation issues ([4f8fe0e](https://github.com/sillydan1/AALTITOAD/commit/4f8fe0e91ce525b2f308f904ccf7f13743565430) by Asger Gitz-Johansen).
- Fix new expr and ctl-expr usages ([228ee15](https://github.com/sillydan1/AALTITOAD/commit/228ee15328fc9eca7535f90c604da1220125de41) by Asger Gitz-Johansen).
- Add huppaal_parser unit tests ([e7d2697](https://github.com/sillydan1/AALTITOAD/commit/e7d26974f297e536eca9ab0ebbaa4efe05ef1187) by Asger Gitz-Johansen).
- Handle edge case when query is reachable after initial tock ([79f5904](https://github.com/sillydan1/AALTITOAD/commit/79f59048ff1f9271588712f61ff2d1dcb9b8cff4) by Asger Gitz-Johansen).
- Forgot about external_variables in operator<< and operator+ for ntta_t ([b1bb2bb](https://github.com/sillydan1/AALTITOAD/commit/b1bb2bbe64c4d5ce87894be23391df1e71d216f4) by Asger Gitz-Johansen).
- Ntta_builder should be able to take external variables ([432ed67](https://github.com/sillydan1/AALTITOAD/commit/432ed6759351e0e0ecef38bcaeba186fcea4c06f) by Asger Gitz-Johansen).
- Forgot to check for satisfaction in a specific edge case ([d60ed02](https://github.com/sillydan1/AALTITOAD/commit/d60ed022242e08a45f8d576d85e6b264b992c1bf) by Asger Gitz-Johansen).
- Add cmake additions of --coverage flags when -dcode_coverage is provided ([b7aeedf](https://github.com/sillydan1/AALTITOAD/commit/b7aeedfa2c150c98442c7c38539d64adde086cc1) by Asger Gitz-Johansen).
- Use the correct uniform distribution in random.cpp ([4fe5124](https://github.com/sillydan1/AALTITOAD/commit/4fe512415f5e40db0d73b2a203bbe3c2e7b7b5b1) by Asger Gitz-Johansen).
- Code coverage setup ([5c53f86](https://github.com/sillydan1/AALTITOAD/commit/5c53f8694d0ee4d38a03ba38d0f682a1dcb62923) by Asger Gitz-Johansen).
- Logic bug where unnecessary states were added to the waiting list ([1e8229e](https://github.com/sillydan1/AALTITOAD/commit/1e8229e4556753518c639154219197c778e042de) by Asger Gitz-Johansen).
- Increase performance for interesting_tocker ([46a54e5](https://github.com/sillydan1/AALTITOAD/commit/46a54e5a6c4312a08e6326c00c7e0ccbc6d22ef8) by Asger Gitz-Johansen).
- Moved ntta_builder implementations into the cpp-file ([22d86b5](https://github.com/sillydan1/AALTITOAD/commit/22d86b510263971b12c9056e6cfb6b08bd2ee8c3) by Asger Gitz-Johansen).
- Component location manipulation was malformed ([57c6f7d](https://github.com/sillydan1/AALTITOAD/commit/57c6f7d957a062605971074dcb5bb5bb3d47d8c2) by Asger Gitz-Johansen).

### Code Refactoring
- Comment cleanup ([dd501f5](https://github.com/sillydan1/AALTITOAD/commit/dd501f54d5339a6f56218dcdaf1074768245eaaa) by Asger Gitz-Johansen).
- Reformulated a todo and cleaned up ([41d2a1e](https://github.com/sillydan1/AALTITOAD/commit/41d2a1e8957cf27d9e8f5ba30b815192b35196aa) by Asger Gitz-Johansen).
- Add some extra tocker information trace logs ([33cef48](https://github.com/sillydan1/AALTITOAD/commit/33cef48f8c7619c000dfee714cab946e1b8662e2) by Asger Gitz-Johansen).

### Features
- Add --no-warn command line option to verifier and simulator ([968dd80](https://github.com/sillydan1/AALTITOAD/commit/968dd80278a2b38a82524a52bb2f6649a4effa4f) by Asger Gitz-Johansen).
- Add query file and direct query to verifier cli ([c8163d3](https://github.com/sillydan1/AALTITOAD/commit/c8163d3e369c71c47f41ebd509cdcc101a497790) by Asger Gitz-Johansen).
- Add verifier cli ([0015d9e](https://github.com/sillydan1/AALTITOAD/commit/0015d9e98d2d1ca6b002c02f2ebc2b97ddc5cee7) by Asger Gitz-Johansen).

src/runtime/TTA.cpp
/**
 * Copyright (C) 2020 Asger Gitz-Johansen

   This file is part of aaltitoad.

    aaltitoad is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    aaltitoad is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with aaltitoad.  If not, see <https://www.gnu.org/licenses/>.
 */
#include <runtime/TTA.h>
#include <extensions/overload>
#include <extensions/cparse_extensions.h>
#include <tinytimer/Timer.hpp>
#include <verifier/trace_output/TTAResugarizer.h>

auto TTA::operator==(const TTA &other) -> bool {
    for(auto& c : components) {
        if(c.second.currentLocation.identifier != other.components.at(c.first).currentLocation.identifier)
            return false;
    }
    for(auto& s : symbols.map()) {
        if(s.second != other.symbols.map()[s.first])
            return false;
    }
    return true;
}

auto TTA::operator!=(const TTA &other) -> bool {
    return !(*this == other);
}

TTA::StateChange operator+(TTA::StateChange a, TTA::StateChange b) {
    // Merge a and b
    // a.symbols.map().merge(b.symbols.map());
    TTA::StateChange cpy{};
    for(auto& s : a.symbols.map())
        cpy.symbols[s.first] = s.second;
    for(auto& s : b.symbols.map())
        cpy.symbols[s.first] = s.second;
    for(auto& c : a.componentLocations)
        cpy.componentLocations[c.first] = c.second;
    for(auto& c : b.componentLocations)
        cpy.componentLocations[c.first] = c.second;
    return cpy;
}

TTA operator<<(const TTA& aa, const TTA::StateChange& b) {
    TTA a{};
    a.components = aa.components;
    a.InsertInternalSymbols(aa.symbols);
    a.InsertExternalSymbols(aa.GetExternalSymbols());
    bool changedSuccessfully = a.SetCurrentState(b);
    if(!changedSuccessfully) spdlog::critical("Something went wrong when trying to apply a statechange.");
    return a;
}

#define FUNC_IMPL(a, func, b, res) std::visit(overload{[&b,&res](auto&& x){std::visit(overload{[&x,&res](auto&& y){res = func(x,y);}}, static_cast<TTASymbol_t>(b));}}, static_cast<TTASymbol_t>(a))
template<typename T1, typename T2>
auto t_ee(const T1&, const T2&) {
    return false;
}
auto t_ee(const bool& a, const bool& b) {return a == b;}
auto t_ee(const int& a, const int& b) {return a == b;}
auto t_ee(const long& a, const long& b) {return a == b;}
auto t_ee(const int& a, const long& b) {return a == b;}
auto t_ee(const long& a, const int& b) {return a == b;}
auto t_ee(const int& a, const float& b) {return a == b;}
auto t_ee(const long& a, const float& b) {return a == b;}
auto t_ee(const float& a, const int& b) {return a == b;}
auto t_ee(const float& a, const long& b) {return a == b;}
auto t_ee(const float& a, const float& b) {return a == b;}
auto t_ee(const TTATimerSymbol& a, const TTATimerSymbol& b) {return a.current_value == b.current_value;}
auto t_ee(const std::string& a, const std::string& b) {return a == b;}
bool operator==(const TTASymbol_t& a, const TTASymbol_t& b) {
    bool retVal = a.index() == b.index();
    if(!retVal)
        return false;
    FUNC_IMPL(a, t_ee, b, retVal);
    return retVal;
}
bool operator!=(const TTASymbol_t& a, const TTASymbol_t& b) {
    return !(a == b);
}

TTASymbol_t TTASymbolValueFromTypeAndValueStrings(const std::string& typestr, const std::string& valuestr) {
    return PopulateValueFromString(TTASymbolTypeFromString(typestr), valuestr);
}

TTASymbol_t TTASymbolTypeFromString(const std::string& typestr) {
    TTASymbol_t value;
    if     (typestr == "int") value = (int)0;
    else if(typestr == "long") value = (long)0;
    else if(typestr == "float") value = (float)0;
    else if(typestr == "bool") value = (bool)false;
    else if(typestr == "string") value = (std::string)"";
    else spdlog::error("Variable type '{0}' is not supported", typestr);
    // Value string
    return value;
}

TTASymbol_t PopulateValueFromString(const TTASymbol_t& type, const std::string& valuestr) {
    TTASymbol_t value = type;
    std::visit(overload(
            [&valuestr, &value](const float&      ){ value = std::stof(valuestr); },
            [&valuestr, &value](const int&        ){ value = std::stoi(valuestr); },
            [&valuestr, &value](const long&       ){ value = std::stol(valuestr); },
            [&valuestr, &value](const bool&       ){
                if(valuestr == "false") value = false;
                else if(valuestr == "true") value = true;
                else spdlog::error("Value '{0}' is not of boolean type", valuestr); // TODO: true/false is case-sensitive. Is this good?
            },
            [&valuestr, &value](const TTATimerSymbol&) { value = TTATimerSymbol{std::stof(valuestr)}; },
            [&valuestr, &value](const std::string&){
                if(valuestr[0] == '\"' && valuestr[valuestr.size()-1] == '\"')
                    value = valuestr.substr(1,valuestr.size()-2);
                else {
                    spdlog::warn("Missing '\"' on string value '{0}' - Adding them manually, but this should be corrected", valuestr);
                    value = valuestr;
                }
            }
    ), value);
    return value;
}

// TODO: This can be optimized by caching hashes on the state changes only. And there's no need to copy the symbols.
std::size_t TTA::GetCurrentStateHash() const {
    std::size_t state_hash = 0;
    for(auto& component : components)
        state_hash == 0 ?
        [&state_hash, &component](){ state_hash = std::hash<std::string>{}(component.second.currentLocation.identifier);}() :
        hash_combine(state_hash, component.second.currentLocation.identifier);

    for(auto& symbol : symbols.map()) {
        auto symbol_hash = std::hash<std::string>{}(symbol.first);   // hash of the symbol identifier
        // Combine with the symbol value
        switch(symbol.second->type) {
            case INT:   hash_combine(symbol_hash, symbol.second.asInt()    * COMBINE_MAGIC_NUM); break;
            case BOOL:  hash_combine(symbol_hash, symbol.second.asBool()   * COMBINE_MAGIC_NUM); break;
            case REAL:  hash_combine(symbol_hash, symbol.second.asDouble() * COMBINE_MAGIC_NUM); break;
            case STR:   hash_combine(symbol_hash, symbol.second.asString()); break;
            case TIMER: hash_combine(symbol_hash, symbol.second.asDouble() * COMBINE_MAGIC_NUM); break;
            default: spdlog::error("Symbol type '{0}' is not supported!", tokenTypeToString(symbol.second->type)); break;
        }
        hash_combine(state_hash, symbol_hash * COMBINE_MAGIC_NUM); // Combine with the overall state
    }
    return state_hash;
}

TTA::ComponentLocationMap TTA::GetCurrentLocations() const {
    // TODO: Maybe this should be cached
    ComponentLocationMap componentLocations = {};
    for(auto& component : components) componentLocations[component.first] = component.second.currentLocation;
    return componentLocations;
}

std::vector<std::string> TTA::GetCurrentLocationsLocationsOnly() const {
    std::vector<std::string> componentLocations = {};
    for(auto& component : components)
        componentLocations.push_back(component.second.currentLocation.identifier);
    return componentLocations;
}

bool TTA::SetCurrentState(const StateChange& newstate) {
    bool result = SetComponentLocations(newstate.componentLocations);
    result &= SetSymbols(newstate.symbols);
    tickCount++;
    return result;
}

bool TTA::SetComponentLocations(const ComponentLocationMap &locationChange) {
    for(auto& componentLocation : locationChange) {
        auto compit = components.find(componentLocation.first);
        if(compit != components.end())
            compit->second.currentLocation = componentLocation.second;
        else {
            spdlog::critical("Attempted to change the state of TTA failed. Component '{0}' does not exist.", componentLocation.first);
            return false;
        }
    }
    return true;
}

bool TTA::SetSymbols(const SymbolMap &symbolChange) {
    for(auto& symbol : symbolChange.map()) {
        auto symbolit = symbols.map().find(symbol.first);
        if(!TypeCheck(symbol, symbolit))
            return false;
        symbols.map()[symbol.first] = symbol.second;
        if(externalSymbols.find(symbol.first) != externalSymbols.end())
            externalSymbols[symbol.first] = symbols.find(symbol.first);
    }
    return true;
}

bool TTA::TypeCheck(const std::pair<const std::string, packToken> &symbol,
                    const std::map<std::string, packToken>::iterator &changingSymbol) const {
    if(changingSymbol == symbols.map().end()) {
        spdlog::critical("Attempted to change the state of TTA failed. Symbol '{0}' does not exist.", symbol.first);
        return false;
    }
    auto x = symbol.second->type;
    auto y = changingSymbol->second->type;
    if((NUM & x & y) || (x == VAR && (NUM & y)) || (x == y))
        return true;
    auto a = tokenTypeToString(changingSymbol->second->type);
    auto b = tokenTypeToString(symbol.second->type);
    spdlog::critical(
            "Attempted to change the state of TTA failed. Symbol '{0}' does not have the correct type. ({1} vs {2} (a := b))",
            symbol.first, a, b);
    return false;
}

bool TTA::IsCurrentStateImmediate() const {
    return std::any_of(components.begin(), components.end(), [](const auto& c){ return c.second.currentLocation.isImmediate; });
}

std::optional<StateMultiChoice> TTA::GetChangesFromEdge(const TTA::Edge& choice, bool& outInfluenceOverlap, component_overlap_t& overlappingComponents, const std::string& currentComponent) const {
    StateMultiChoice changes{};
    changes.component = currentComponent;
    bool DoesUpdateInfluenceOverlap = AccumulateUpdateInfluences(choice, changes.symbolsToChange, overlappingComponents, currentComponent);
    outInfluenceOverlap |= DoesUpdateInfluenceOverlap;
    if(DoesUpdateInfluenceOverlap)
        return {};

    for(auto& symbolChange : changes.symbolsToChange)
        changes.symbolChanges.push_back(symbolChange.second);

    return changes;
}

void TTA::WarnAboutComponentOverlap(component_overlap_t& overlappingComponents) const {
    std::vector<std::string> overlapping_errors{};
    for(auto it = overlappingComponents.begin(), end = overlappingComponents.end(); it != end; it = overlappingComponents.equal_range(it->first).second) {
       auto iter = overlappingComponents.equal_range(it->first);
       const std::string& compName = iter.first->second.componentName;
       for(auto& x = iter.first; x != iter.second; x++) {
           if(x->second.componentName != compName) {
               overlapping_errors.emplace_back(x->second.componentName + " and " + compName + " disagree on: " + it->first);
               break;
           }
       }
    }
    if(overlapping_errors.empty())
        return;

    spdlog::trace("Overlapping Updates:");
    for(auto& err : overlapping_errors)
        spdlog::trace(err);
}

TokenMap TTA::GetSymbolChangesAsMap(std::vector<UpdateExpression> &symbolChanges) const {
    SymbolMap symbolsCopy{};
    for(auto& symbolChange : symbolChanges) {
        if(symbols.map()[symbolChange.lhs]->type == TIMER)
            symbolsCopy[symbolChange.lhs] = packToken(symbolChange.Evaluate(symbols).asDouble(), PACK_IS_TIMER);
        else
            symbolsCopy[symbolChange.lhs] = symbolChange.Evaluate(symbols);
    }
    return symbolsCopy;
}

void TTA::ApplyComponentLocation(TTA::ComponentLocationMap &currentLocations,
                                 const std::pair<const std::string, TTA::Component> &component,
                                 TTA::Edge &pickedEdge) {// If we transition to the end location, immediately change to first location
    if(component.second.endLocation.identifier == pickedEdge.targetLocation.identifier)
        currentLocations[component.first] = component.second.initialLocation;
    else
        currentLocations[component.first] = pickedEdge.targetLocation;
}

std::vector<TTA::StateChange> TTA::GetNextTickStates(const nondeterminism_strategy_t& strategy) const {
    // Result type: [0] is shared, [>0] are choice changes
    StateMultiChoice sharedChanges{};
    std::vector<StateMultiChoice> choiceChanges{};

    component_overlap_t overlappingComponents{}; // expr.lhs -> componentIdentifiers mapping
    bool updateInfluenceOverlapGlobal = false;
    for(auto& component : components) {
        auto enabledEdges = component.second.GetEnabledEdges(symbols); // TODO: Take Interesting variables into account
        if(!enabledEdges.empty()) {
            bool hasNondeterminism = WarnIfNondeterminism(enabledEdges, component.first);
            for(auto& edge : enabledEdges) {
                auto changes = GetChangesFromEdge(edge, updateInfluenceOverlapGlobal, overlappingComponents, component.first);
                if(changes.has_value()) {
                    ApplyComponentLocation(changes->currentLocations, component, edge);
                    choiceChanges.push_back(changes.value());
                }
            }
        }
    }
    if(updateInfluenceOverlapGlobal)
        WarnAboutComponentOverlap(overlappingComponents);
    auto symbolsCopy = GetSymbolChangesAsMap(sharedChanges.symbolChanges);
    std::vector<TTA::StateChange> retVal{{ sharedChanges.currentLocations, symbolsCopy }};
    for(auto& change : choiceChanges)
        retVal.push_back({change.currentLocations, GetSymbolChangesAsMap(change.symbolChanges)});
    return retVal;
}

bool TTA::WarnIfNondeterminism(const std::vector<Edge>& edges, const std::string& componentName) {
    // If we aren't simulating, we should not warn about nondeterminism
    // TODO: This has been disabled, because it destroys non-trace proofs!
    // if(!CLIConfig::getInstance()["trace"]) return false;
    if(edges.size() > 1) {
        spdlog::debug("Non-deterministic choice in Component '{0}'.", TTAResugarizer::Resugar(componentName));
        spdlog::debug("Enabled edges in component '{0}':", TTAResugarizer::Resugar(componentName));
        for(auto& e : edges)
            spdlog::debug("{0} --> {1}",
                          TTAResugarizer::Resugar(e.sourceLocation.identifier),
                          TTAResugarizer::Resugar(e.targetLocation.identifier));
        spdlog::debug("----- / -----");
        return true;
    }
    return false;
}

std::string TTA::GetCurrentStateString() const {
    std::stringstream ss{}; ss << "{";
    for(auto& component : components)
        ss<<"\""<<component.first<<"\""<<": "<<"\""<<component.second.currentLocation.identifier<<"\",";
    for(auto& symbol : symbols.map()) {
        if(symbol.second->type == TIMER)
            ss << "\"" << symbol.first << "\"" << ": " << "\"" << symbol.second.asDouble() << "\",";
        else
            ss << "\"" << symbol.first << "\"" << ": " << "\"" << symbol.second.str() << "\",";
    }
    ss << R"("OBJECT_END":"true"})"; // This is just a bad way of ending a json object. 
    return ss.str();
}

std::vector<TTA::Edge> TTA::Component::GetEnabledEdges(const SymbolMap& symbolMap) const {
    auto edges_from_current_state = edges.equal_range(currentLocation.identifier);
    std::vector<TTA::Edge> ret{};
    for(auto edge = edges_from_current_state.first; edge != edges_from_current_state.second; ++edge) {
        if(edge->second.guardExpression.empty()) {
            ret.push_back(edge->second);
            continue; // Empty guards are considered to be always satisfied
        }
        auto res = calculator::calculate(edge->second.guardExpression.c_str(), symbolMap);
        if(res.asBool()) ret.push_back(edge->second);
    }
    return ret;
}

void TTA::Tick(const nondeterminism_strategy_t& nondeterminismStrategy) {
    Timer<int> timer;
    timer.start();
    SetCurrentState(GetNextTickStates(nondeterminismStrategy)[0]);
    spdlog::info("Tick {0} time elapsed: {1} ms - (With printing and everything)", tickCount, timer.milliseconds_elapsed());
    tickCount++;
}

void TTA::InsertExternalSymbols(const TTA::SymbolMap& externalSymbolKeys) {
    symbols.map().insert(externalSymbolKeys.map().begin(), externalSymbolKeys.map().end());
    for(auto& elem : externalSymbolKeys.map())
        externalSymbols[elem.first] = symbols.find(elem.first);
}

void TTA::InsertInternalSymbols(const TTA::SymbolMap &internalSymbols) const {
    GetSymbols().map().insert(internalSymbols.map().begin(), internalSymbols.map().end());
}

std::optional<const TTA::Component*> TTA::GetComponent(const std::string &componentName) const {
    auto it = components.find(componentName);
    if(it == components.end()) return {};
    return { &it->second };
}

TTA::TTA() {
    // TODO: This is a stupid hack. Please implement an extension for cparse, so that the "False" literal is not case sensitive.
    // TODO: This hack makes it possible to assign "false := true", which is insanity incarnate.
    symbols.map()["false"] = false;
    symbols.map()["true"] = true;
}

TTA::Edge& TTA::PickEdge(std::vector<TTA::Edge>& edges, const nondeterminism_strategy_t &strategy) const {
    if(edges.empty()) throw std::logic_error("Trying to pick an edge from an empty list of edges is impossible.");
    if(edges.size() == 1) return edges[0];
    switch (strategy) {
        case nondeterminism_strategy_t::PICK_FIRST:
            return edges[0];
        case nondeterminism_strategy_t::PICK_LAST:
            return edges[edges.size()-1];
        case nondeterminism_strategy_t::PICK_RANDOM:
            return edges[rand() % edges.size()];
        default:
        case nondeterminism_strategy_t::PANIC:
            spdlog::critical("Panicking due to nondeterministic choice!");
            throw std::exception();
    }
}

void TTA::DelayAllTimers(double delayDelta) {
    for(auto& symbol : symbols.map()) {
        if(symbol.second->type == TIMER)
            symbols[symbol.first] = packToken(static_cast<double>(symbol.second.asDouble() + delayDelta), PACK_IS_TIMER);
    }
}

[[maybe_unused]] void TTA::SetAllTimers(double exactTime) {
    for(auto& symbol : symbols.map()) {
        if(symbol.second->type == TIMER)
            symbols[symbol.first] = packToken(exactTime, PACK_IS_TIMER);
    }
}

void TTA::StateChange::DelayTimerSymbols(SymbolMap& symbols, float delayDelta) {
    for(auto& symbol : symbols.map()) {
        if(symbol.second->type == TIMER)
            symbols[symbol.first] = packToken(static_cast<double>(symbol.second.asDouble() + delayDelta), PACK_IS_TIMER);
    }
}

bool AreRightHandSidesIndempotent(const std::map<std::string, std::vector<std::pair<std::string,std::string>>>& overlappingComponents, const std::string& key, const TTA::SymbolMap& symbols) {
    auto xx = overlappingComponents.find(key);
    bool indempotence = true;
    auto origval = calculator::calculate(xx->second.begin()->second.c_str(), symbols);
    for(auto& x : xx->second) {
        auto val = calculator::calculate(x.second.c_str(), symbols);
        if(origval != val)
            return false;
    }
    return indempotence;
}

bool TTA::AccumulateUpdateInfluences(const TTA::Edge& pickedEdge, std::multimap<std::string, UpdateExpression>& symbolsToChange, component_overlap_t& overlappingComponents, const std::string& currentComponent) const {
    bool updateInfluenceOverlap = false;
    for(auto& expr : pickedEdge.updateExpressions) {
        bool overlap_in_this_expr = false;
        auto it = overlappingComponents.find(expr.lhs);
        if(it != overlappingComponents.end()) {
            auto range = overlappingComponents.equal_range(expr.lhs);
            for(auto iter = range.first; iter != range.second; iter++) {
                if(iter->second.componentName == currentComponent)
                    continue;
                // TODO: Check for Idempotence!
                spdlog::debug("Overlapping update influence on evaluation of update on edge {0} --> {1}. "
                              "Variable '{2}' is already being written to in this tick! - For more info, run with higher verbosity",
                              TTAResugarizer::Resugar(pickedEdge.sourceLocation.identifier),
                              TTAResugarizer::Resugar(pickedEdge.targetLocation.identifier),
                              TTAResugarizer::Resugar(expr.lhs)); // TODO: Idempotent variable assignment
                updateInfluenceOverlap = true;
                overlap_in_this_expr = true;
            }
        }
        component_overlap c{
                .componentName=currentComponent,
                .updates={std::make_pair(TTAResugarizer::Resugar(
                        expr.lhs + " : " +
                        pickedEdge.sourceLocation.identifier + " --> " +
                        pickedEdge.targetLocation.identifier), expr.rhs)}
        };
        overlappingComponents.emplace(expr.lhs, std::move(c));
        if(overlap_in_this_expr)
            continue;
        symbolsToChange.insert({ expr.lhs, expr });
    }
    return updateInfluenceOverlap;
}

bool TTA::IsDeadlocked() const {
    return std::all_of(components.begin(), components.end(), [&] (const auto& component) {
        auto enabledEdges = component.second.GetEnabledEdges(symbols);
        return enabledEdges.empty();
    });
}

bool TTA::IsSymbolExternal(const std::string &identifier) const {
    return externalSymbols.find(identifier) != externalSymbols.end();
}

std::vector<TTA::Edge> TTA::GetCurrentEdges() const {
    std::vector<Edge> edges{};
    for(auto& component : components) {
        auto currentEdgesInComponent = component.second.edges.equal_range(component.second.currentLocation.identifier);
        for(auto edge = currentEdgesInComponent.first; edge != currentEdgesInComponent.second; ++edge) {
            edges.push_back(edge->second);
        }
    }
    return edges;
}

void TTA::InsertExternalSymbols(const TTA::ExternalSymbolMap &externalSymbolKeys) {
    for(auto& elem : externalSymbolKeys) {
        externalSymbols[elem.first] = symbols.find(elem.first);
    }
}

bool TTA::StateChange::IsEmpty() const {
    return componentLocations.empty() && symbols.map().empty();
}
